File: app/src/main/java/com/example/kriptorep4ik/logic/JsoupParser.java
package com.example.kriptorep4ik.logic;

import android.util.Log;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class JsoupParser {
    private final ExecutorService executor = Executors.newSingleThreadExecutor();                   //фоновый поток

    public void init() {
        executor.execute(this::getWeb);                                                             //"вызови метод getWeb() в фоновом потоке"
    }

    private void getWeb() {
        try {
            Document doc = Jsoup.connect("https://www.cbr.ru/currency_base/daily/").get();
            Elements tables = doc.getElementsByTag("tbody");
            Element our_table = tables.get(0);
            Elements elements_from_table = our_table.children();
            Element AUD = elements_from_table.get(1);

            Log.d("MyLog", "Tbody size: " + AUD.text());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/com/example/kriptorep4ik/logic/JsoupParser.kt
package com.example.kriptorep4ik.logic

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements

class JsoupParser {

    suspend fun getWeb(): List<CurrencyRateModel> {
        return withContext(Dispatchers.IO) {

            try {
                val doc: Document = Jsoup.connect("https://www.cbr.ru/currency_base/daily/").get()
                val tables: Elements = doc.getElementsByTag("tr")
                val currencyRatesContainer = mutableListOf<CurrencyRateModel>()

                for (i in 1 until  tables.size) {
                    val our_table: Element = tables[i]
                    val temp: Elements = our_table.children()

                    val letter_code: Element = temp[1]
                    val rate: Element = temp[temp.size - 1]

                    val currency: String = if (temp.size == 6) {
                        "${temp[3].text()} ${temp[4].text()}"

                    } else {

                        temp[3].text()
                    }
                    val currencyRateModel = CurrencyRateModel(
                        currency = currency,
                        letter_code = letter_code.text(),
                        rate = rate.text()
                    )
                    currencyRatesContainer.add(currencyRateModel)

                    Log.d("mylog", our_table.text())
                    Log.d("mylog", letter_code.text())
                    Log.d("mylog", currency)
                    Log.d("mylog", rate.text())

                }

                currencyRatesContainer

            } catch (e: Exception) {
                Log.e("mylog", "Error in getWeb: ${e.message}", e)
                emptyList()
            }
        }
    }
}